lib/http/response-validator/validationStreams.js

```diff
@@ -0,0 +1,131 @@
+var AWS = require('../../core');
+var cryptoLib = AWS.util.crypto.lib;
+var Buffer = AWS.util.Buffer;
+var TransformStream = AWS.util.stream.Transform;
+var CHECKSUM_BYTE_LEN = 16;
+
+module.exports = {
+  IntegrityCheckerStream: IntegrityCheckerStream,
+  ContentLengthCheckerStream: ContentLengthCheckerStream,
+}
+
+require('util').inherits(IntegrityCheckerStream, TransformStream);
+require('util').inherits(ContentLengthCheckerStream, TransformStream);
+
+/**
+ * Only runnable in Node
+ * A transform stream to perform trailing md5 hash validation.
+ * Whether to perform this validation is based on 'x-amz-tranfer-encoding'
+ * response header.
+ * @api private
+ */
+function IntegrityCheckerStream(transferEncoding, contentLength) {
+  if (!(this instanceof TransformStream)) {
+    return new TransformStream();
+  }
+  TransformStream.call(this);
+  if (transferEncoding !== 'append-md5') {
+    process.nextTick(function(stream, encoding) {
+      stream.emit(
+        'error',
+        AWS.util.error(
+          new Error('Cannot validate response header: ' + encoding + ', expect \'append-md5\''),
+          {code: 'UnsupportedHashingAlgorithm'}
+        )
+      );
+    }, this, transferEncoding);
+  }
+  this.hash = cryptoLib.createHash('md5');
+  this.providedChecksum = typeof Buffer.allocUnsafe === 'function' ?
+    Buffer.allocUnsafe(CHECKSUM_BYTE_LEN).fill(0) :
+    new Buffer(CHECKSUM_BYTE_LEN);
+  this.providedChecksumLen = 0;
+  this.expectedDataLength = contentLength - CHECKSUM_BYTE_LEN;
+  this.receivedLen = 0
+}
+
+/**
+ * Override the flush function in Stream.Transform. It check whether the hashing is correct before
+ * 'end' event emitted and after all data has been transformed and output.
+ */
+IntegrityCheckerStream.prototype._flush = function(callback) {
+  if (
+    this.providedChecksum.length !== CHECKSUM_BYTE_LEN ||
+    this.hash.digest('base64') !== this.providedChecksum.toString('base64')
+  ) {
+    this.emit('error', AWS.util.error(
+      new Error('Response fails integrity check.'),
+      { code: 'ResponseChecksumMismatch' })
+    );
+  }
+  callback();
```

Comment:
Can you just call the `callback` with the error object? I think that takes care of emitting the error if you do.